src/main/java/lc/kra/system/keyboard/GlobalKeyboardHook.java

```diff
@@ -132,19 +138,66 @@ public GlobalKeyboardHook(boolean raw) throws UnsatisfiedLinkError {
 	 * @param event A global key event
 	 */
 	private void keyPressed(GlobalKeyEvent event) {
+		
+		int virtualKeyCode = event.getVirtualKeyCode();
+		
+		// only add if key is not already hold down
+		if(!isKeyHoldDown(event))
+			holdDownKeyCodes.add(virtualKeyCode);
+		
 		for(GlobalKeyListener listener:listeners)
 			listener.keyPressed(event);
 	}
+	
 	/**
 	 * Invoke keyReleased (transition state TS_UP) for all registered listeners
 	 * 
 	 * @param event A global key event
 	 */
 	private void keyReleased(GlobalKeyEvent event) {
+
+		int virtualKeyCode = event.getVirtualKeyCode();
+
+		// redundant if - key should always be hold down if you can release it, but you never know
+		if(isKeyHoldDown(event))
+			holdDownKeyCodes.remove(virtualKeyCode);
+		
 		for(GlobalKeyListener listener:listeners)
 			listener.keyReleased(event);
 	}
 
+	/**
+	 * Checks if the specified key is currently hold down
+	 * 
+	 * @param virtualKeyCode the virtual code of the key, use constants in {@link GlobalKeyEvent}
+	 * 
+	 * @return true if the key is currently hold down
+	 */
+	public boolean isKeyHoldDown(int virtualKeyCode) {
+		return holdDownKeyCodes.contains(virtualKeyCode);
+	}
+	
+	
+	/**
+	 * Checks if all the specified keys are currently hold down
+	 * 
+	 * @param virtualKeyCodes any number of specified key codes, use constants in {@link GlobalKeyEvent}
+	 * 
+	 * @return true if all the specified keys are currently hold down, false if any of the keys is not currently hold down
+	 */
+	public boolean areKeysHoldDown(int... virtualKeyCodes) {
+		for(int keyCode : virtualKeyCodes) {
+			if(!isKeyHoldDown(keyCode)) {
+				return false;
+			}
+		}
+		return true;
+	}
+	
+	private boolean isKeyHoldDown(GlobalKeyEvent event) {
```

Comment:
No longer necesarry, when comment in the previous line is updated.

Reply:
done, removed